Remove decorators and use webpack to build {N} app with AoT

NathanWalker opened the pull request:
@m-abs @d3viant0ne let's continue here.
Working webpack build, just a runtime error:
```
file:///app/bundle.js:37207:49: JS ERROR TypeError: moduleFactory.create is not a function. (In 'moduleFactory.create(ngZoneInjector)', 'moduleFactory.create' is undefined)
```

m-abs committed:
chor: {N}-app is now build with nativescript-dev-webpack

Changes:
* nativescript-dev-typescript have been removed
* nativescript-dev-webpack have been installed
* {N} sourcecode have been moved from nativescript/app
  to nativescript/src, including App_Resources and symlinks
* The nativescript/app/assets symlink have been removed, these files are
  copied by gulp-tasks.
* To build the {N}-app you now have to run:
  gulp build.tns && cd nativescript && tns build android --bundle

Gulp tasks:
* gulp clean.tns:
  * Empties nativescript/app
* gulp build.tns_html_css:
  * Copies the *.tns.html/css-files to nativescript/app and removed
    'tns' from their filename. This is to remove the need for @BaseComponent
* gulp build.assets.tns:
  * Copies files from src/client/assets to nativescript/app/assets
* gulp build.js.tns:
  * Gulp transpiles TypeScript from nativescript/src to nativescript/app
    Note: tsconfig.json have been moved to nativescript/src/tsconfig.json
* gulp watch.tns:
  * Watch for changes for the {N}-app
* gulp build.tns:
  * Groups: build.tns_html_css, build.assets.tns, build.js.tns
* gulp build.tns.watch:
  * Watches with watch.tns and runs build.tns on changes

m-abs committed:
fix: sass build for {N}

m-abs committed:
fix: fonts weren't copied

m-abs committed:
fix: for tns-related tasks and start adding build configs for tns

* Adds tslint.tns task
* Updates npm tasks for building/running {N}-app
* Fixes watches
* Fixes update html-files after watch
* Fixes for livesync

m-abs committed:
update watchlist

m-abs committed:
Merge remote-tracking branch 'upstream/master' into playing-with-build-3

m-abs committed:
chor: update npm scripts

m-abs committed:
chor: EnvConfig in {N}-app

m-abs committed:
chor: adds build.prod.tns

NathanWalker committed:
chore: merge upstream parent

NathanWalker committed:
cleanup

NathanWalker committed:
feat: improve sass setup with {N} app

m-abs committed:
chor: add npm run nativescript-install to all tns-related start.* scripts

m-abs committed:
Merge remote-tracking branch 'upstream/master' into new-build-system

m-abs committed:
fix: do 'npm run nativescript-install' before gulp build.tns

m-abs committed:
fix: start.ios and start.android

m-abs committed:
chor: callenableProdMode for prod in {N}-app

m-abs committed:
fix: Removed HomeComponent.scrollToItem

m-abs committed:
fix: shouldn't have removed preinstall hook from nativescript/package.json

m-abs committed:
Merge remote-tracking branch 'upstream/master' into new-build-system

m-abs committed:
chor: update to new nativescript-dev-webpack

NathanWalker committed:
chore: merge upstream parent

NathanWalker committed:
fix: injectables

NathanWalker committed:
test wip

NathanWalker committed:
feat: webpack building complete - still run issue

NathanWalker committed:
feat: add vendor

NathanWalker committed:
feat(AoT): {N} webpack working build, runtime error wip

m-abs commented on nativescript/webpack.common.js:
```diff
@@ -0,0 +1,171 @@
+        // Compile TypeScript files with ahead-of-time compiler.
+        {
+          test: /\.ts$/,
+          loaders: [
```
@NathanWalker 
Why did you change all but this one from `loaders` to `use`?

NathanWalker replied:
thanks forgot that one, updated now.

m-abs commented on nativescript/webpack.common.js:
```diff
@@ -0,0 +1,171 @@
+      //Angular AOT compiler
+      new AotPlugin({
+        tsConfigPath: 'tsconfig.aot.json',
+        entryModule: 'app/native.module#NativeModule',
```
I'm wondering if the *.aot.ts should have the same basename as entryModule?
The current aot.ts-file is named app.aot.ts, maybe it should be named native.module.aot.ts?

NathanWalker replied:
I don't think that matters here as it's the app.ts file which is the `main` targeted from `package`.

m-abs commented on nativescript/webpack.config.js:
```diff
@@ -0,0 +1,64 @@
+        },
+        to: '[path]/[name]'
+      },
+      // @nota plugins
```
Wupsy. this shouldn't have been added to the seed project. :-1: 
This is one of the modules, we use in LYT3, the project I'm working on.

m-abs commented on nativescript/package.json:
```diff
@@ -1,58 +1,83 @@
+    "main": ".tmp/app.js",
+    "version": "0.0.0",
+    "nativescript": {
+        "id": "com.yourdomain.nativescript",
```
This doesn't match the name in the `app.gradle`.

NathanWalker replied:
I have updated `app.gradle` to match.

m-abs commented on tools/config/seed-advanced.config.ts:
```diff
@@ -1,11 +1,30 @@
 
   constructor() {
     super();
+    this.ENABLE_SCSS = true;
```
Should SCSS be enabled? I think this line got copied by accident from our project.

m-abs commented on src/client/app/frameworks/sample/sample.module.ts:
```diff
@@ -28,6 +28,10 @@ import { MultilingualModule } from '../i18n/multilingual.module';
   ],
+  schemas: [
+    NO_ERRORS_SCHEMA,
+    CUSTOM_ELEMENTS_SCHEMA
```
I understand `NO_ERRORS_SCHEMA`, but why `CUSTOM_ELEMENTS_SCHEMA`?

NathanWalker replied:
http://stackoverflow.com/questions/39428132/custom-elements-schema-added-to-ngmodule-schemas-still-showing-error

m-abs commented on src/client/app/frameworks/core/services/app.service.ts:
```diff
@@ -0,0 +1,14 @@
+import { MultilingualService } from '../../../frameworks/i18n/index';
+
+@Injectable()
+export class AppService {
```
Just to make it clear, since I haven't explained it.

`AppService` and `NSAppService` were created because `nativescript-dev-webpack` didn't like the `NSAppComponent` that inherited from `AppComponent` but wasn't real a `@Component`.
It's only supported to get injected in the `AppComponent`, the idea is to setup `NativeScript` listeners and more.

NathanWalker replied:
Love it, great idea 👍

m-abs commented on src/client/app/components/home/home.component.ts:
```diff
@@ -1,21 +1,28 @@
 export class HomeComponent {
   public names$: Observable<any>;
   public newName: string = '';
+  private _cnt: number = 3; // just for example purposes
```
Leftover from something? It's not used.

m-abs commented on nativescript/src/shared/core/services/ns-app.service.ts:
```diff
@@ -0,0 +1,166 @@
+export class NSAppService extends AppService {
+  nsApp = nsApp;
+
+  externalRouteMap = [
```
Wupsy: More leftover from our project.
This is meant for deeplink we use.

m-abs committed:
stuff

m-abs committed:
Merge remote-tracking branch 'upstream/new-build-system-new-webpack' into new-build-system-new-webpack

joshwiens commented on nativescript/src/_common.scss:
```diff
@@ -0,0 +1,26 @@
+
+// TextField {
+//   margin: 15 15 0 15;
+// }
```
Pull the commented out block

joshwiens commented on nativescript/src/components.module.ts:
```diff
@@ -4,16 +4,16 @@ import { NativeScriptModule, NativeScriptFormsModule, NativeScriptHttpModule, Na
 // app
-import { WindowService, ConsoleService, RouterExtensions } from './app/frameworks/core/index';
-import { NSAppComponent } from './pages/app/app.component';
+import { WindowService, ConsoleService, RouterExtensions, AppService } from './app/frameworks/core/index';
```
```
import { 
  WindowService, 
  ConsoleService, 
  RouterExtensions, 
  AppService 
} from './app/frameworks/core/index';
```

For anything with more than three imports, particularly with something with a long path multi-line imports tend to be a bit more readable imo.

joshwiens commented on nativescript/src/native.module.ts:
```diff
@@ -0,0 +1,58 @@
+// nativescript
+import { NativeScriptModule, NativeScriptFormsModule, NativeScriptHttpModule, NativeScriptRouterModule, RouterExtensions as TNSRouterExtensions } from 'nativescript-angular';
```
Same as above, I would multi-line this one as well

m-abs committed:
chor: renamed app.aot.ts => native.module.aot.ts moved app config to native.module.ts

joshwiens commented on nativescript/src/tns-java-classes.ts:
```diff
@@ -0,0 +1,25 @@
+//`require` calls get resolved to the correct bundling import call.
+//
+//At runtime the module gets loaded *before* the rest of the app code, so code
+//placed here needs to be careful about its dependencies.
```
Would multi-line jsdoc style comments be a bit more readable than 9 single line comments?

m-abs replied:
This file should be removed, it's from the old-version of `nativescript-dev-webpack`

m-abs replied:
In newer versions the same functionality is covered by vendor.ts and vendor-platform.android/ios.ts, they look just as bad, but they're from a template so I don't think we should alter them too much.

joshwiens commented on nativescript/src/app.aot.ts:
```diff
@@ -0,0 +1,6 @@
+// this import should be first in order to load some required settings (like globals and reflect-metadata)
+import { platformNativeScript } from "nativescript-angular/platform-static";
```
This will fail with the current tslint settings

```
    "quotemark": [
      true,
      "single"
    ],
```

joshwiens commented on nativescript/src/tns-java-classes.ts:
```diff
@@ -0,0 +1,25 @@
+            return require("ui/frame/activity");
+        }
+    };
+}
```
This will fail with the current tslint settings

```
    "quotemark": [
      true,
      "single"
    ],
```

joshwiens commented on nativescript/src/vendor-platform.android.ts:
```diff
@@ -0,0 +1,25 @@
+            return require("ui/frame/activity");
+        }
+    };
+}
```
This will fail with the current tslint settings

```
    "quotemark": [
      true,
      "single"
    ],
```

joshwiens commented on tools/tasks/seed/build.js.tns.ts:
```diff
@@ -0,0 +1,70 @@
+        .pipe(gulp.dest(Config.TNS_APP_DEST));
+
+      const copy = projectFiles()
+        //.pipe(plugins.template(template))
```
Should be pulled if we aren't using it.

m-abs replied:
It was commented out because the template plugin threw an error because of it.
It should either get fixed or better yet implemented in webpack.

joshwiens replied:
I'd vote for just moving that into Webpack as well.

m-abs commented:
Could we do something about node-sass that keeps failing in ci?

NathanWalker committed:
feat(NativeScript): AoT webpack bundling wip

NathanWalker committed:
feat(NativeScript): AoT webpack bundling wip

NathanWalker committed:
feat(NativeScript): AoT webpack bundling wip

NathanWalker committed:
wip

NathanWalker commented:
I'm not concerned about CI until we can at least get everything else working.

NathanWalker committed:
wip

joshwiens commented:
But how do we know it's working without CI :p

m-abs committed:
fix: *sic* don't mess with the indentation

m-abs committed:
Merge remote-tracking branch 'upstream/new-build-system-new-webpack' into new-build-system-new-webpack

m-abs committed:
chor: lock to the same dep versions as hallo world.

* Locked webpack to 2.1.0-beta.27 as it gives another error than 2.1.0-beta.28

m-abs commented:
I discovered that I could break my hallo world project by upgrading to angular 2.2.3 from 2.2.1. So I downgraded this project to 2.2.1. I also discovered that I had a new version og webpack in this project than in my hallo world, so I downgraded that aswell.

Now I get a new Error when I try to run this project.

```
Cannot redefine property: Config in....
```

I'm looking into what happens here, but it looks like progress from the NativeModule is not an NgModule error I had before.

m-abs commented:
Once I solved that problem, I got the NgModule error again...

m-abs committed:
fix: Fix some AoT errors.

m-abs committed:
chor: undo renaming of app.ts

m-abs commented:
I fixed most of the errors I got from running ngc manually.
It complains about paths to includes in the generated ngfactory.ts-files, but I think this might be the reason for the webpack-plugin?
```
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/app/components/about/about.component.ngfactory.ts:22:27: Cannot find module '../../../../ui/page'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/app/components/app.component.ngfactory.ts:31:27: Cannot find module '../../../ui/page'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/app/components/app.component.ngfactory.ts:39:27: Cannot find module '../../../ui/frame'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/native.module.ngfactory.ts:69:27: Cannot find module '../ui/frame'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/native.module.ngfactory.ts:76:27: Cannot find module '../platform'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/native.module.ngfactory.ts:103:27: Cannot find module '../ui/page'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/native.module.ngfactory.ts:104:27: Cannot find module '../src/translate.service'.
```

The gulp build.tns-task didn't copy the aot.ts file into the app folder, so webpack didn't attempt to build it...
Unfortunately I now get this error from webpack:
```
ERROR in ./app.aot.ts                                                                                                                                                                                                                  
 Module build failed: Error: /home/mabs/project/angular2-seed-advanced/nativescript/app/app.aot.ts (4,39): Cannot find module './native.module.ngfactory'.)                                                                                
     at _transpile (/home/mabs/project/angular2-seed-advanced/nativescript/node_modules/@ngtools/webpack/src/loader.js:101:19)                                                                                                           
     at /home/mabs/project/angular2-seed-advanced/nativescript/node_modules/@ngtools/webpack/src/loader.js:128:26                                                                                                                        
     at process._tickCallback (internal/process/next_tick.js:103:7)                                                                                                                                                                      
```

m-abs committed:
chor: android bundle with webpack now runs

m-abs commented:
Hi @NathanWalker 

It took a lot of trail and error, but now I can both build and run with webpack bundles on this branch (at least on Android, I hope iOS works as well)

I had to move TranslateModule.forRoot(...) to `ComponentModule` from your `MultilingualModule` and I rolled `tns-core-modules` back to `2.4.3` from `next` and add `baseUrl` to the `tsconfig.aot.json`.
The problems I had building `app.aot.ts` were caused by strange import paths generated by `ngc`, especially for `ng2-translate`.

We still need one major thing before this can be merged. Template variables must be ported from the gulp tasks to this one.

I also stopped using the `tns-loader`, since it have no real function as long as the gulp task copies/renames the template files. I left in git, since we are going to need it.

It's my hope when `nativescript-dev-webpack` supports livesync, we'd no longer need the `src`-folder and we no longer need the gulp tns-tasks.

@NathanWalker could you merge master into this branch when you have some time for it? I could resolve the conflict or I might just be too tired. :)

NathanWalker commented:
@m-abs that is very exciting news!

I will get latest merged up to this branch early next week, thanks 👍

NathanWalker committed:
chore: merge master

NathanWalker committed:
chore: merge master

m-abs committed:
fix: make tslint happy again

m-abs committed:
Merge branch 'new-build-system-new-webpack' of github.com:NathanWalker/angular2-seed-advanced into new-build-system-new-webpack

m-abs committed:
chor: typescript@^2.1.0 in root project

m-abs commented:
Hi again,

I'm getting some conflicts in the root project, rxjs was updated to @5.0.2 but both ng2-config and angular@2.2.3 wants a peerDep of rxjs@5.0.0_beta12.
*sic* Why are npm deps so troublesome?

I managed to fix most of the problems (so I at least could install root-project) by downgrading to typescript@2.1.4 (the first version to fix the symlink issue). I don't know what this will do to web aot builds...

m-abs committed:
fix: ng2-config broke webpack build

m-abs committed:
fix: make both web and tns run

vip-git commented:
a quick question @NathanWalker and @m-abs :
- Webpack build is currently targeted towards nativescript or will this work on web build as well ?

I was planning to send a PR for webpack build for the webclient but if this is already taken care off then this is awesome !!

m-abs committed:
chor: adds template variable injection to tns webpack

m-abs commented:
@vip-git 
This PR only deals with nativescript.

@NathanWalker 
My last commit adds support for template variables in webpack and reenabled the `tns-loader`, which now have a purpose :)

I have to generate a JSON-file in the `gulp build.js.tns` task, otherwise I don't know how to extract the values from the `typescript`-files under `tools` and get them into webpack for tns.

Do you have a better suggestion? Something that'd work for both web/desktop and for tns?

m-abs committed:
chor: remove hack for typescript 2.2_dev in gulp tslint

m-abs committed:
fix: indentation of tsconfig.json

m-abs commented:
@NathanWalker 
Another issue (I keep running into new ones)
The `@Effects` decorator doesn't work with `AOT`.
Where is a workaround, which I'll look at tomorrow: https://github.com/ngrx/effects/pull/79

NathanWalker commented:
@m-abs thank you for all the investigations. I must say `@Effects` decorator is a surprise, I would expect that to 'just work' but apparently everything with AoT is a chance...ergh.

NathanWalker committed:
chore: merge master

m-abs committed:
chor: NSLocationStrategy._getSatates renamed to NSLocationStrategy._getStates()

m-abs committed:
chor: update angular to 2.3.1 for {N}

m-abs committed:
chor: Update {N} to angular-2.3.1

m-abs committed:
Merge branch 'new-build-system-new-webpack' of https://github.com/NathanWalker/angular2-seed-advanced into new-build-system-new-webpack

m-abs commented:
We have to wait until `nativescript-angular` have been updated to angular-2.4.x+, AOT and `@ngrx/effect` breaks with angular-2.3.x, right now the bundled App cannot start up due to a DI issue. 

I'll be working on other tasks for the next week or so.

m-abs committed:
Merge branch 'master' of github.com:NathanWalker/angular2-seed-advanced into new-build-system-new-webpack

m-abs committed:
chor: update deps

m-abs commented:
I've run into a new problem: https://github.com/NativeScript/nativescript-dev-webpack/issues/54

ericdoerheit committed:
Update build.js.dev.ts

Correcting PROJECT_ROOT and APP_SRC references for optional debugging with Webstorm/IntelliJ

mgechev committed:
Merge pull request #1767 from ericdoerheit/patch-1

Update build.js.dev.ts

admosity committed:
feature(new): User defined watch paths

admosity committed:
feat: long zone.js stack traces for dev

mgechev committed:
Merge pull request #1769 from admosity/zone-long-stack-trace

feat: zone.js long stack traces for dev

mgechev committed:
Merge pull request #1768 from admosity/extra-paths-watch

feat: User defined watch paths

m-abs committed:
chor: Make bundling work with ng2.4

m-abs committed:
Merge remote-tracking branch 'upstream/master' into new-build-system-new-webpack

m-abs commented:
I have bundling up and running now. We still need to test it properly.

NathanWalker commented:
👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏  @m-abs bravo sir!
I'll take a look - would be amazing to get this merged this week. Thank you.

NathanWalker commented:
@m-abs can you provide here the latest workflow with the latest changes?
For running development and when doing a release?

mgechev committed:
docs: update the list of contributors

mgechev committed:
docs: update the list of contributors

m-abs commented:
***Development:***
`npm run start.livesync.android`
`npm run start.livesync.ios`

***Testing bundles:***
gulp build.prod.tns  && cd nativescript && npm run start-android-bundle
gulp build.prod.tns && cd nativescript && npm run start-ios-bundle
(Should be added as npm scripts)

***Publishing (Not tested):***
**iOS:**
gulp build.prod.tns --env-config prod --build-type prod && cd nativescript && tns publish ios --bundle --team-id <TEAM-ID>

**Android:**
gulp build.prod.tns --env-config prod --build-type prod && cd nativescript && tns build android --bundle --key-store-path <KEY-STORE-PATH> --key-store-password <KEY-STORE-PASS> --key-store-alias <KEY-STORE-ALIAS> --key-store-alias-password <KEY-STORE-ALIAS-PASSWORD>  --copy-to release.apk
Upload `release.apk` to Google Play.

In our project we have some scripts/tasks that handles building and publishing releases for both platforms.
I intend to clean these up and contribute them.

m-abs commented:
@NathanWalker 
There is a few things we need before this is merged.

* Document: Service replaced with DI have to adhere to the API of the original class. This gave me a lot of problems.
* Document: Components/Directives/Pipes have to belong to a single `NgModule` or `ngc` will fail. This means an unused component have to be deleted and a component used in more than one `NgModule` have to be put in a single NgModule that the others can include.
* Document: The new dev, test, release workflow.
* Add: npm-scripts for building and releasing bundles.
* SCSS? Right now, because we use it in our project, the {N}-app uses SCSS but the angular-seed-project this is supposed to be optional.

m-abs commented:
Then I tried to integrate this in our own project, ngc kept complaining about components/pipes inside `nativescript/src` not being in an NgModule, even though `nativescript/src` is in the `tsconfig.aot.json`'s exclude list.

I solved it by moving the directive in `CORE_DIRECTIVES` in to their own `NgModule` in `src/client/app/frameworks/core/directives/index.ts`.

I'm wondering if this is a good idea to do upstream?
Although I couldn't reproduce it with this project.

It seems to work fine with out own project, on my phone startup time have gone from 20+ seconds to ~2seconds.

NathanWalker commented:
@m-abs 
* Let's enable SCSS by default (I'd consider that to be a desired out of box config for an advanced project like this) - I've always enabled on the 3 projects I have started myself based off this seed.
* Regarding directive in `CORE_DIRECTIVES` into own NgModule - *YES* let's move it here in the seed.
* `Service replaced with DI have to adhere to the API of the original class. This gave me a lot of problems.` - what are you referring to here?

NathanWalker committed:
feat: enhanced action bar util to setup iOS and Android - also ensured AppService gets constructed

joshwiens commented:
`Service replaced with DI have to adhere to the API of the original class. This gave me a lot of problems` 

Curious about that myself

NathanWalker committed:
cleanup injection

NathanWalker committed:
feat: use mobile device default language

m-abs commented:
> Service replaced with DI have to adhere to the API of the original class. This gave me a lot of problems

`NSAngulartics2Segment` is injected to replace `Angulartics2Segment` in {N}, which made `ngc` complain about the `NSAngulartics2Segment`.
If I remember correctly, `ngc` complained that `Angulartics2Segment` had a private `angulartics2`, while `NSAngulartics2Segment` didn't have one.

mgechev committed:
chore: remove duplicate files

NathanWalker committed:
gitignore

NathanWalker committed:
chore: merge upstream

NathanWalker committed:
feat: restructing - webpack loader for NativeScript views, no more custom decorators, AoT builds for web and mobile

NathanWalker committed:
feat: restructing - webpack loader for NativeScript views, no more custom decorators, AoT builds for web and mobile
